Scaler AWSHelperFn should not override list context

The below test illustrates this issue. The param property is defined as a list of basestrings. This should then allow a list of AWSHelperFn as well (param=[Ref(fake)] below). However, a scaler AWSHelperFn should cause an error (param=Ref(fake) below) but is accepted. Correct behavior would be to raise a TypeError exception.
```
$ cat helper_test.py
import unittest

from troposphere import AWSObject, AWSProperty, Ref

class FakeRef(AWSObject):
    props = {
    }

class Prop(AWSProperty):
    props = {
        'param': ([basestring], True),
    }

class TestHelper(unittest.TestCase):
    def test_helper(self):
	fake = FakeRef('fakeref')
        Prop(param=["fakeref"])
        with self.assertRaises(TypeError):
            Prop(param="fakeref")
        Prop(param=[Ref(fake)])
        with self.assertRaises(TypeError):
            Prop(param=Ref(fake))

if __name__ == '__main__':
    unittest.main()

$ python helper_test.py
F
======================================================================
FAIL: test_helper (__main__.TestHelper)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "helper_test.py", line 22, in test_helper
    Prop(param=Ref(fake))
AssertionError: TypeError not raised

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
```

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -379,5 +379,26 @@ class TestJoin(unittest.TestCase):
             Join(10, "foobar")
 
 
+class TestHelperFn(unittest.TestCase):
+
+    def test_helperfn(self):
+        class FakeRef(AWSObject):
+            props = {
+            }
+
+        class Prop(AWSProperty):
+            props = {
+                'param': ([basestring], True),
+            }
+
+        fake = FakeRef('fakeref')
+        Prop(param=["fakeref"])
+        with self.assertRaises(TypeError):
+            Prop(param="fakeref")
+        Prop(param=[Ref(fake)])
+        with self.assertRaises(TypeError):
+            Prop(param=Ref(fake))
+
+
 if __name__ == '__main__':
     unittest.main()

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -388,13 +388,13 @@ class TestHelperFn(unittest.TestCase):
 
         class Prop(AWSProperty):
             props = {
-                'param': ([basestring], True),
+                'param': ([int], True),
             }
 
         fake = FakeRef('fakeref')
-        Prop(param=["fakeref"])
+        Prop(param=[1])
         with self.assertRaises(TypeError):
-            Prop(param="fakeref")
+            Prop(param=1)
         Prop(param=[Ref(fake)])
         with self.assertRaises(TypeError):
             Prop(param=Ref(fake))

--- a/troposphere/__init__.py
+++ b/troposphere/__init__.py
@@ -152,8 +152,11 @@ class BaseAWSObject(object):
 
             # If the value is a AWSHelperFn we can't do much validation
             # we'll have to leave that to Amazon.  Maybe there's another way
-            # to deal with this that we'll come up with eventually
-            if isinstance(value, AWSHelperFn):
+            # to deal with this that we'll come up with eventually.
+            # We'll do validation below if there are AWSHelperFn in a
+            # list context.
+            if isinstance(value, AWSHelperFn) and \
+                    not isinstance(expected_type, list):
                 return self.properties.__setitem__(name, value)
 
             # If it's a function, call it...

